Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp

```diff
@@ -0,0 +1,255 @@
+        return integer.divided_by("10"_bigint).remainder;
+    };
+
+    auto res = m_numerator.divided_by(m_denominator);
+    BigFraction tmp { move(res.quotient) };
+
+    auto const needed_power = Crypto::NumberTheory::Power("10"_bigint, Crypto::UnsignedBigInteger { rounding_threshold });
+    // We get one more digit to do proper rounding
+    auto const frac_value = res.remainder.multiplied_by(needed_power.multiplied_by("10"_bigint)).divided_by(m_denominator).quotient;
+
+    tmp.m_numerator.set_to(tmp.m_numerator.multiplied_by(needed_power));
+    tmp.m_numerator.set_to(tmp.m_numerator.plus(frac_value.divided_by("10"_bigint).quotient));
+    if (get_last_digit(frac_value) > "4"_bigint)
+        tmp.m_numerator.set_to(tmp.m_numerator.plus("1"_bigint));
+
+    tmp.m_denominator.set_to(tmp.m_denominator.multiplied_by(needed_power));
+
+    return tmp;
+}
+
+void BigFraction::reduce()
+{
+    auto const GCD = Crypto::NumberTheory::GCD(m_numerator.unsigned_value(), m_denominator);
+
+    if (GCD == 1)
+        return;
+
+    auto const numerator_divide = m_numerator.divided_by(GCD);
+    VERIFY(numerator_divide.remainder == "0"_bigint);
+    m_numerator = numerator_divide.quotient;
+
+    auto const denominator_divide = m_denominator.divided_by(GCD);
+    VERIFY(denominator_divide.remainder == "0"_bigint);
+    m_denominator = denominator_divide.quotient;
+}
+
+String BigFraction::to_string(unsigned rounding_threshold) const
+{
+    StringBuilder builder;
+    if (m_numerator.is_negative() && m_numerator != "0"_bigint)
+        builder.append("-");
+
+    auto const number_of_digits = [](auto integer) {
+        unsigned size = 1;
+        for (auto division_result = integer.divided_by(Crypto::UnsignedBigInteger { 10 });
+             division_result.remainder == Crypto::UnsignedBigInteger { 0 } && division_result.quotient != Crypto::UnsignedBigInteger { 0 };
+             division_result = division_result.quotient.divided_by(Crypto::UnsignedBigInteger { 10 })) {
+            ++size;
+        }
+        return size;
+    };
+
+    auto const rounded_fraction = rounded(rounding_threshold);
+
+    // We take the unsigned value as we already manage the '-'
+    auto const full_value = rounded_fraction.m_numerator.unsigned_value().to_base(10);
+    int split = full_value.length() - (number_of_digits(rounded_fraction.m_denominator) - 1);
+
+    if (split < 0)
+        split = 0;
+
+    auto const remove_tailing_zeros = [](StringView value) -> StringView {
+        auto n = value.length();
+        while (value.characters_without_null_termination()[n - 1] == '0')
+            --n;
+        return { value.characters_without_null_termination(), n };
+    };
+
+    auto const raw_frac_value = full_value.substring(split, full_value.length() - split);
+
+    String const integer_value = split == 0 ? "0" : full_value.substring(0, split);
+    String const frac_value = rounding_threshold == 0 ? "0" : remove_tailing_zeros(raw_frac_value);
```

Comment:
I don't think these need to be strings at all
```suggestion
    auto const integer_value = split == 0 ? "0"sv : full_value.substring_view(0, split);
    auto const frac_value = rounding_threshold == 0 ? "0"sv : remove_tailing_zeros(raw_frac_value);
```